src/lib.rs

```diff
@@ -0,0 +1,318 @@
+use std::ffi::{CStr, CString};
+use std::fmt::{Display, Formatter};
+use std::os::raw::c_char;
+use std::path::PathBuf;
+use std::sync::{Mutex, MutexGuard};
+
+#[cfg_attr(
+    all(target_os = "windows", target_env = "msvc"),
+    link(name = "golassie.dll")
+)]
+#[cfg_attr(
+    not(all(target_os = "windows", target_env = "msvc")),
+    link(name = "golassie")
+)]
+extern "C" {
+    fn InitDaemon(config: *const GoDaemonConfig) -> InitDaemonResult;
+    fn DropDaemonInitResult(result: *mut InitDaemonResult);
+    fn RunDaemon() -> LassieResult;
+    fn StopDaemon() -> LassieResult;
+    fn DropResult(value: *mut LassieResult);
+}
+
+#[repr(C)]
+#[derive(Debug)]
+struct InitDaemonResult {
+    port: u16,
+    error: *const c_char,
+}
+
+impl Drop for InitDaemonResult {
+    fn drop(&mut self) {
+        unsafe { DropDaemonInitResult(self) }
+    }
+}
+
+impl InitDaemonResult {
+    fn error(&self) -> Option<String> {
+        from_c_string(self.error)
+    }
+}
+
+#[repr(C)]
+#[derive(Debug)]
+struct LassieResult {
+    error: *const c_char,
+}
+
+impl Drop for LassieResult {
+    fn drop(&mut self) {
+        unsafe { DropResult(self) }
+    }
+}
+
+impl LassieResult {
+    fn error(&self) -> Option<String> {
+        from_c_string(self.error)
+    }
+}
+
+fn from_c_string(str: *const c_char) -> Option<String> {
+    if str.is_null() {
+        return None;
+    }
+
+    Some(unsafe { CStr::from_ptr(str) }.to_string_lossy().to_string())
+}
+
+#[allow(dead_code)]
```

Comment:
which dead code is here?

Reply:
None at the moment. In the past, I was not accessing some fields, e.g. `temp_dir`, and the compiler complained that we had dead code. 

I am going to remove this flag, we can easily add it later if needed.